### DISCUSSION TITLE
AVX2 instructions in bottled binaries?

### DISCUSSION THREAD

**Discussion by jeroen** [Homebrew/discussions#620] [Linux] [answered]
I am using linuxbrew to build some portable linux binaries. Some users on older hardware have complained that the binaries are crashing on their machine due to invalid instructions.

Somebody pointed out that the binaries contain `vextracti128` which is an AVX2 instruction. I traced this back to the openssl bottle in linuxbrew:

```
brew install openssl
objdump -d /home/linuxbrew/.linuxbrew/lib/libcrypto.a  | grep vextracti128
```

Linuxbrew binaries are supposed to be built with `-mtune=core2` which should not enable AVX2 instructions? Or am I completely diagnosing the problem incorrectly now.

**Comment by SMillerDev** [comment]
@Homebrew/linux any ideas how this leaked in there?

**Comment by jonchang** [comment, marked as answer]
The presence of AVX instructions in an object file is no guarantee that those instructions will actually be executed. For example, openssl can switch between AVX and non-AVX code paths based on runtime detection of CPU capabilities (or from setting certain environment variables), which is pretty reliable as long as your OS is accurately reporting its features. I'd start looking there to narrow down the possibilities.

**Reply by jonchang** [reply]
Given that these are Linux machines it's also just occurred to me that certain AMD processors might not implement all Intel "core2" instructions, or it's also possible that your users are running Pentiums. If that's the case they'll likely have to build all formulae from source.

**Reply by jeroen** [reply]
I don't think they're _that_ old :) I must have messed up somewhere when building a library without `--build-bottle` such that it was compiled with `-mtune=native`. But it is surprisingly difficult to debug this. Here is the issue: https://github.com/r-lib/gert/issues/117

**Reply by jonchang** [reply]
Yeah, you'd need to pass `--build-bottle` if you're redistributing binaries on Linux.

On macOS we default to using the oldest architecture supported on your OS version when building from source. On Linux this isn't the case and we use `-march=native` when building from source since there isn't really the same concept of support as there is on macOS. This leads to many cases where users will compile software on their login nodes and hit illegal instructions when queueing their jobs because the cluster is full of AMD k8 machines or whatever.

Based on the thread it looks like the new builds have solved your users' issues so hopefully this helped. Runtime probing of CPU features makes it difficult to poke at a binary and guess "will this crash on another machine" without actually running that binary. It's not helped by the fact that there have been bugs with openssl's mechanisms before which confuses the issue even further.

**Reply by jeroen** [reply]
> On macOS we default to using the oldest architecture supported on your OS version when building from source.

This is an interesting tangent, can you point me to where this happens? We have been seeing crashes when running homebrew big-sur binaries in the rosetta emulator on M1 chips, which I suspect, are because homebrew enables hardware extensions that [rosetta is not able to emulate](https://developer.apple.com/documentation/apple_silicon/about_the_rosetta_translation_environment).

**Reply by carlocab** [reply]
Related: https://github.com/Homebrew/homebrew-core/issues/67713

>switch between AVX and non-AVX code paths based on runtime detection of CPU capabilities

That's assuming the compiler isn't capricious enough to optimise those checks away. It may be worth trying a build with minimal optimization flags to see if that changes anything. Or, maybe even an older compiler that hasn't been taught those optimisations yet.
